core/trino-spi/src/main/java/io/trino/spi/predicate/SortedRangeSet.java

```diff
@@ -957,6 +957,44 @@ public long getRetainedSizeInBytes()
         return Optional.of(Collections.unmodifiableList(result));
     }
 
+    /// Return smaller representation of this a discrete value set.
+    @Override
+    public ValueSet compact()
+    {
+        if (!isDiscreteSet()) {
+            return this; // TODO: support also non-discrete sets
+        }
+        Type type = getType();
+        Ranges ranges = getRanges();
+        List<Range> result = new ArrayList<>();
+
+        Optional<Range> current = Optional.empty();
+        for (Range inputRange : ranges.getOrderedRanges()) {
+            Object value = inputRange.getSingleValue();
+            if (current.isPresent()) {
+                Range range = current.get();
+                if (areConsecutive(type, range.getHighBoundedValue(), value)) {
+                    // extend current range to include the new value
+                    current = Optional.of(Range.range(type, range.getLowBoundedValue(), true, value, true));
+                    continue;
+                }
+                // not consecutive values - flush the current range, and start collecting a new one
+                result.add(current.get());
+            }
+            current = Optional.of(Range.equal(type, value));
+        }
+        current.ifPresent(result::add);  // flush last range (if exists)
+        return ValueSet.ofRanges(result);
+    }
+
+    private static boolean areConsecutive(Type type, Object low, Object high)
+    {
+        return type
+                .getDiscreteValues(new Type.Range(low, high))
```

Comment:
Can we use comparison operator to detect adjacent values the way `tryMergeWithNext` does ?
That would avoid relying on `getDiscreteValues` which isn't implemented for many types.

Reply:
Not sure - IIUC, `tryMergeWithNext()` relies on the ranges sharing the low/high bound values, i.e. `[1,5]` and `(5,9]` can be merged into `[1,9]`.
However, if we try to merge `[1,5]` and `[6,9]`, we should make sure that 5 and 6 are "consecutive" (i.e. there is no other value between them, and this requires calling `Type#getDiscreteValues`).

Reply:
Yes, we can't use similar logic as `tryMergeWithNext` here.
But re-using `Type#getDiscreteValues` also seems awkward.
I'm thinking we should just add a `boolean areConsecutive(Object low, Object high)` to Type for this.
@findepi @martint thoughts ?

Reply:
While working on Iceberg I had a need to find an adjacent value (not check whether two values are adjacent).
Here is my API proposal for that: https://github.com/trinodb/trino/pull/12797

Here, it seems we don't need new methods on Type, so I'd hold off and not add them.